autoemulate/experimental/tuner.py

```diff
@@ -94,27 +97,50 @@ def run(  # noqa: PLR0913
         model_config_tested: list[ModelConfig] = []
         val_scores: list[float] = []
 
-        for i in range(self.n_iter):
-            logger.debug(
-                "Tuning Model: Iteration %s: %d/%d",
-                model_class.__name__,
-                i + 1,
-                self.n_iter,
-            )
+        i, retries = 0, 0
+        while i < self.n_iter:
             # randomly sample hyperparameters and instantiate model
             model_config = model_class.get_random_config()
+            try:
+                scores = cross_validate(
+                    cv=KFold(n_splits=n_splits, random_state=seed, shuffle=shuffle),
+                    dataset=self.dataset,
+                    x_transforms=x_transforms,
+                    y_transforms=y_transforms,
+                    model=model_class,
+                    model_config=model_config,
+                    device=self.device,
+                    random_seed=None,
+                )
+                retries = 0
+                model_config_tested.append(model_config)
+                val_scores.append(scores["r2"])  # type: ignore  # noqa: PGH003
+                i += 1
+                logger.debug(
+                    "tuning model: %s; iteration: %d/%d; mean (std) r2=%.3f (%.3f); "
+                    "model_config: %s",
+                    model_class.model_name(),
+                    i + 1,
+                    self.n_iter,
+                    np.mean(scores["r2"]),
+                    np.std(scores["r2"]),
+                    str(model_config),
+                )
 
-            scores = cross_validate(
-                cv=KFold(n_splits=n_splits, random_state=seed, shuffle=shuffle),
-                dataset=self.dataset,
-                x_transforms=x_transforms,
-                y_transforms=y_transforms,
-                model=model_class,
-                model_config=model_config,
-                device=self.device,
-                random_seed=None,
-            )
-            model_config_tested.append(model_config)
-            val_scores.append(scores["r2"])  # type: ignore  # noqa: PGH003
+            except Exception as e:
+                # Retry with new random config parameters
+                retries += 1
+                logger.warning(
+                    "Failed tuning iteration %d with model config: %s: %s",
+                    i + 1,
+                    json.dumps(model_config, default=str, separators=(",", ":")),
+                    str(e),
+                )
+                # If many retries, log error and break
+                if retries > max_retries:
```

Comment:
It feels to me like you'd never get here since the `while` is over `self.n_iter` not `max_retries`

Reply:
E.g. that following the `max_retries` and `attempts` in compare loop it would run https://github.com/alan-turing-institute/autoemulate/blob/5922eed740560aa8393ad54729fd69541a9f02dd/autoemulate/experimental/compare.py#L324-L329 and then move on to the next model.